workers/unity/Packages/com.improbable.gdk.guns/Systems/ServerShootingSystem.cs

```diff
@@ -1,59 +1,59 @@
     {
-        private struct PlayersShooting
+        private WorkerSystem workerSystem;
+        private CommandSystem commandSystem;
+        private ComponentUpdateSystem componentUpdateSystem;
+
+        protected override void OnCreateManager()
         {
-            public readonly int Length;
-            public ComponentDataArray<HealthComponent.CommandSenders.ModifyHealth> ModifyHealthCommandSenders;
-            [ReadOnly] public ComponentDataArray<SpatialEntityId> EntityId;
-            [ReadOnly] public ComponentDataArray<ShootingComponent.ReceivedEvents.Shots> Shots;
-            [ReadOnly] public ComponentDataArray<GunComponent.Component> Gun;
-        }
+            base.OnCreateManager();
 
-        [Inject] private PlayersShooting playersShooting;
+            workerSystem = World.GetExistingManager<WorkerSystem>();
+            commandSystem = World.GetExistingManager<CommandSystem>();
+            componentUpdateSystem = World.GetExistingManager<ComponentUpdateSystem>();
+        }
 
         protected override void OnUpdate()
         {
-            for (var i = 0; i < playersShooting.Length; i++)
-            {
-                var commandSender = playersShooting.ModifyHealthCommandSenders[i];
-                var commandSent = false;
-                var gunId = playersShooting.Gun[i].GunId;
-                var gunSettings = GunDictionary.Get(gunId);
-                var damage = gunSettings.ShotDamage;
+            var events = componentUpdateSystem.GetEventsReceived<ShootingComponent.Shots.Event>();
+            var gunDataForEntity = GetComponentDataFromEntity<GunComponent.Component>();
 
-                foreach (var shot in playersShooting.Shots[i].Events)
+            for (var i = 0; i < events.Count; ++i)
+            {
+                ref readonly var shotEvent = ref events[i];
+                var shotInfo = shotEvent.Event.Payload;
+                if (!ValidateShot(shotInfo))
                 {
-                    var shotInfo = shot;
-                    if (!ValidateShot(shotInfo))
-                    {
-                        continue;
-                    }
-
-                    // Send command to entity being shot.
-                    var modifyHealthRequest = new HealthComponent.ModifyHealth.Request(
-                        new EntityId(shotInfo.EntityId),
-                        new HealthModifier()
-                        {
-                            Amount = -damage,
-                            Origin = shotInfo.HitOrigin,
-                            AppliedLocation = shotInfo.HitLocation
-                        });
-                    commandSender.RequestsToSend.Add(modifyHealthRequest);
-                    commandSent = true;
+                    continue;
                 }
 
-                if (commandSent)
+                var shooterSpatialID = new EntityId(shotInfo.EntityId);
+                if (!workerSystem.TryGetEntity(shooterSpatialID, out var shooterEntity))
```

Comment:
Does it matter which order `ValidateShot` and this check are done in? 

If not - we can put the less expensive one first